### DISCUSSION TITLE
simulations running very slow

### DISCUSSION THREAD

**Discussion by ashishdhole** [idaholab/moose#28612] [Q&A Modules: Phase field]
Hello,
I am running a phase field simulation and the input file looks like this 
```
  [Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 260
  ny = 260
  xmin = 0
  xmax = 2000
  ymin = 0
  ymax = 2000
  uniform_refine = 2
[]
[GlobalParams]
  op_num = 17
  var_name_base = gr
  length_scale = 1.0e-6
  time_scale = 1.0
[]

[Variables]
  [PolycrystalVariables]
    order = FIRST
    family = LAGRANGE
  []
[]
[Modules]
  [AdvanceFrictionPressureGG]
    g = phi
  []
[]

[UserObjects]
  [voronoi]
    type = PolycrystalVoronoi
    grain_num = 6400 # Number of grains
    rand_seed = 2878
    coloring_algorithm = jp
    int_width = 5
  []
  [grain_tracker]
    type = GrainTracker
    verbosity_level = 0
    compute_var_to_feature_map = true
  []
[]
[ICs]
  [PolycrystalICs]
    [PolycrystalColoringIC]
      polycrystal_ic_uo = voronoi
    []
  []
[]
.
.
.
.
.
[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = PJFNK #Preconditioned JFNK (default)
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
  steady_state_detection = true
  steady_state_start_time = 0
  steady_state_tolerance = 1e-8
  automatic_scaling = true
  compute_scaling_once = false
  use_pre_SMO_residual = false
  l_tol = 1.0e-4 # Linear relative tolerance
  l_abs_tol = 1.0e-6 # Linear absolute tolerance
  nl_abs_tol = 1.0e-4 # Nonlinear absolute tolerance
  nl_rel_tol = 1.0e-4 # Nonlinear relative tolerance
  l_max_its = 100 # Linear max iterations
  nl_max_its = 30 # Nonlinear max iterations
  start_time = 0.0
  #num_steps = 10
  end_time = 5000000000
  dtmin = 1e-20
  #dt = 1
  #line_search = none
  #[Adaptivity]
  #  initial_adaptivity = 3
  #  refine_fraction = 0.7
  #  coarsen_fraction = 0.4
  #  max_h_level = 2
  #[]
  [TimeStepper]
    type = IterationAdaptiveDT
    cutback_factor = 0.8
    dt = 0.1
    growth_factor = 1.2
  []
[]

[Adaptivity]
  initial_steps = 3
  max_h_level = 2
  marker = err_bnds
  [Markers]
    [err_bnds]
      type = ErrorFractionMarker
      coarsen = 0.4
      refine = 0.7
      indicator = ind_bnds
    []
  []
  [Indicators]
    [ind_bnds]
      type = GradientJumpIndicator
      variable = bnds
    []
  []
[]
```
but my simulation is running very slow can you suggest what am I I doing wrong?
output looks like this
```
Setting Up
  Uniformly Refining...                                                                  [^[[33m 31.72 s^[[39m] [^[[33m  909 MB^[[39m]
  Initializing
    Initializing Equation Systems......                                                  [^[[33m 43.78 s^[[39m] [^[[33m 1563 MB^[[39m]
  Finished Initializing                                                                  [^[[33m 47.48 s^[[39m] [^[[33m 1564 MB^[[39m]
Finished Setting Up                                                                      [^[[33m 83.60 s^[[39m] [^[[33m 1564 MB^[[39m]
Framework Information:
MOOSE Version:           git commit 8b30259223 on 2024-04-17
LibMesh Version:         7dbd90ade88cb16a5e855a4c345454c9c146ebec
PETSc Version:           3.20.3
SLEPc Version:           3.20.1
Current Time:            Thu Sep 12 06:25:21 2024
Executable Timestamp:    Thu Sep 12 02:51:06 2024

Parallelism:
  Num Processors:          300
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:	   2
  Nodes:
    Total:                 1083681
    Local:                 3757
    Min/Max/Avg:           3440/3769/3612
  Elems:
    Total:                 1081600
    Local:                 3625
    Min/Max/Avg:           3551/3639/3605
  Num Subdomains:          1
  Num Partitions:          300
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                18422577
  Num Local DOFs:          63869
  Num Constrained DOFs:    35377
  Local Constrained DOFs:  0
  Variables:               { "gr0" "gr1" "gr2" "gr3" "gr4" ... "gr12" "gr13" "gr14" "gr15" "gr16" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                6493762
  Num Local DOFs:          22014
  Num Constrained DOFs:    4162
  Local Constrained DOFs:  0
  Variables:               "bnds" { "unique_grains" "var_indices" } "phi" { "err_bnds" "ind_bnds" }
  Finite Element Types:    "LAGRANGE" "MONOMIAL" "LAGRANGE" "MONOMIAL"
  Approximation Orders:    "FIRST" "CONSTANT" "FIRST" "CONSTANT"
^[[31mLEGACY MODES ENABLED:^[[39m
 This application uses the legacy initial residual evaluation behavior. The legacy behavior performs an often times redundant residual evaluation before the solution modifying objects are executed prior to the initial (0th nonlinear iter$
^[[39m
Currently Executing
  Performing Initial Setup
    Computing User Objects
      Computing Polycrystal Initial Condition.                                           [^[[33m 18.03 s^[[39m] [^[[33m 1610 MB^[[39m]
      Currently Finalizing Polycrystal Initial Condition
        Finalizing Polycrystal Initial Condition
          Communicating and Merging                                                      [^[[33m  5.08 s^[[39m] [^[[33m 1643 MB^[[39m]
        Finished Finalizing Polycrystal Initial Condition                                [^[[33m  5.16 s^[[39m] [^[[33m 1644 MB^[[39m]
      Finished Finalizing Polycrystal Initial Condition                                  [^[[33m  6.22 s^[[39m] [^[[33m 1972 MB^[[39m]
    Finished Computing User Objects                                                      [^[[33m 24.25 s^[[39m] [^[[33m 1972 MB^[[39m]
    Projecting Initial Solutions........................................................................................................................................................... [^[[33m785.21 s^[[39m] [^[[33m 1972 MB^[[39m]
Mesh unchanged, skipping remaining steps...
    Finished Building SemiLocalElemMap                                                   [^[[33m  6.75 s^[[39m] [^[[33m 2021 MB^[[39m]
  Finished Performing Initial Setup                                                      [^[[33m827.94 s^[[39m] [^[[33m 2022 MB^[[39m]
  Outputting
    Outputting Step.                                                                     [^[[33m 15.11 s^[[39m] [^[[33m 2257 MB^[[39m]

Time Step 0, time = 0
Postprocessor Values:
+----------------+----------------+----------------+----------------+
| time           | avg_grain_vol  | g              | grain_tracker  |
+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   5.917303e+02 |   0.000000e+00 |   6.400000e+03 |
+----------------+----------------+----------------+----------------+

  Finished Outputting                                                                    [^[[33m 15.14 s^[[39m] [^[[33m 2257 MB^[[39m]

Time Step 1, time = 0.1, dt = 0.1
Pre-SMO residual: 222.403

Performing automatic scaling calculation

 0 Nonlinear |R| = ^[[32m9.331901e+00^[[39m
      0 Linear |R| = ^[[32m9.331901e+00^[[39m
      1 Linear |R| = ^[[32m6.595481e-01^[[39m
      2 Linear |R| = ^[[32m9.307747e-02^[[39m
      3 Linear |R| = ^[[32m2.443747e-02^[[39m
      4 Linear |R| = ^[[32m5.012671e-03^[[39m
      5 Linear |R| = ^[[32m1.524739e-03^[[39m
      6 Linear |R| = ^[[32m4.031388e-04^[[39m
 1 Nonlinear |R| = ^[[32m8.499808e-02^[[39m
      0 Linear |R| = ^[[32m8.499808e-02^[[39m
      1 Linear |R| = ^[[32m1.393766e-02^[[39m
      2 Linear |R| = ^[[32m1.382264e-03^[[39m
      3 Linear |R| = ^[[32m1.702454e-04^[[39m
      4 Linear |R| = ^[[32m3.164961e-05^[[39m
      5 Linear |R| = ^[[32m5.682206e-06^[[39m
 2 Nonlinear |R| = ^[[32m1.111908e-02^[[39m
      0 Linear |R| = ^[[32m1.111908e-02^[[39m
      1 Linear |R| = ^[[32m5.624584e-04^[[39m
      2 Linear |R| = ^[[32m9.147334e-05^[[39m
      3 Linear |R| = ^[[32m1.074289e-05^[[39m
      4 Linear |R| = ^[[32m3.286875e-06^[[39m
      5 Linear |R| = ^[[32m5.419093e-07^[[39m
 3 Nonlinear |R| = ^[[32m3.724282e-04^[[39m
^[[32m Solve Converged!^[[39m
  Finished Solving                                                                       [^[[33m 26.03 s^[[39m] [^[[33m 2294 MB^[[39m]

Outlier Variable Residual Norms:
  gr5: ^[[33m1.607098e-04^[[39m
  gr7: ^[[33m1.786166e-04^[[39m

Postprocessor Values:
+----------------+----------------+----------------+----------------+
| time           | avg_grain_vol  | g              | grain_tracker  |
+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   5.917303e+02 |   0.000000e+00 |   6.400000e+03 |
|   1.000000e-01 |   5.925049e+02 |   1.000000e-01 |   6.400000e+03 |
+----------------+----------------+----------------+----------------+
```
thanks
Ashish

**Comment by GiudGiud** [comment]
Hello

How did you install moose? Which installation route?
And where are you running this? I see 300 processes

**Reply by ashishdhole** [reply]
I am running it on HPC.

**Reply by ashishdhole** [reply]
The standard procedure that is shown on the website.

**Reply by ashishdhole** [reply]
After 5 hours I could only go till timestep 
Time Step 63, time = 94.4506, dt = 3.68568

that is really slow

**Reply by GiudGiud** [reply]
which HPC?
which page did you follow on the website?

**Reply by ashishdhole** [reply]
Its my institute's cluster. And I followed
https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html

**Reply by GiudGiud** [reply]
ok good. 
You can try tuning the hypre preconditioning parameters through the petsc options

**Reply by ashishdhole** [reply]
any suggestions? 
As there are so many options. I had tried many so far like lu ilu but still its very slow.

**Reply by GiudGiud** [reply]
hypre should be good
some info there
https://mooseframework.inl.gov/releases/moose/v1.0.0/application_development/hypre.html

**Reply by ashishdhole** [reply]
Thanks. I will check it out and get back to you.

**Comment by ashishdhole** [comment]
Hello 
I am using following solver for the system posted at the beginning of this discussion thread.
```
  type = Transient
  scheme = bdf2
  solve_type = PJFNK #Preconditioned JFNK (default)
  petsc_options_iname = '-pc_type -pc_hypre_type -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_agg_nl -pc_hypre_boomeramg_agg_num_paths  -pc_hypre_boomeramg_max_levels  -pc_hypre_boomeramg_coarsen_type  -pc_hypre_boomeramg_interp_type  -pc_hypre_boomeramg_P_max  -pc_hypre_boomeramg_truncfactor'
  petsc_options_value = 'hypre boomeramg  0.7 4 5 25 HMIS ext+i 2 0.3'
```
for some reason, my simulations are still very slow. I am using 300 cores and yet the simulations are very slow. Adapting mesh after every time step is taking really long time. more time than actual iterations. I don't think a 2d simulations with 6400 grains takes 2 days. If I increase the mobility then my time step increment reduces. If I try to coarsen the mesh the results do not have enough resolution. Please help.
Thank you

**Reply by GiudGiud** [reply]
I can't remember, are you using the grain tracker? And how high of an order parameter ?

**Reply by ashishdhole** [reply]
yes.
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 250
  ny = 250
  xmin = 0
  xmax = 2000
  ymin = 0
  ymax = 2000
  uniform_refine = 2
[]
[GlobalParams]
  op_num = 17
  var_name_base = gr
[]

[Variables]
  [PolycrystalVariables]
    order = FIRST
    family = LAGRANGE
  []
[]
[Modules]
  [AdvanceFrictionPressureGG]
    g = phi
  []
[]

[UserObjects]
  [voronoi]
    type = PolycrystalVoronoi
    grain_num = 6400 # Number of grains
    rand_seed = 2878
    coloring_algorithm = jp
    int_width = 3.46
  []
  [grain_tracker]
    type = GrainTracker
    verbosity_level = 0
    compute_var_to_feature_map = true
  []
[]
[ICs]
  [PolycrystalICs]
    [PolycrystalColoringIC]
      polycrystal_ic_uo = voronoi
    []
  []
[]
```

**Comment by ashishdhole** [comment]
Simulation environment details
```
Setting Up
  Uniformly Refining...                                                                  [^[[33m 28.40 s^[[39m] [^[[33m  863 MB^[[39m]
  Initializing
    Initializing Equation Systems......                                                  [^[[33m 41.54 s^[[39m] [^[[33m 1473 MB^[[39m]
  Finished Initializing                                                                  [^[[33m 45.25 s^[[39m] [^[[33m 1474 MB^[[39m]
Finished Setting Up                                                                      [^[[33m 79.66 s^[[39m] [^[[33m 1474 MB^[[39m]
Framework Information:
MOOSE Version:           git commit 8b30259223 on 2024-04-17
LibMesh Version:         7dbd90ade88cb16a5e855a4c345454c9c146ebec
PETSc Version:           3.20.3
SLEPc Version:           3.20.1
Current Time:            Fri Sep 20 21:16:02 2024
Executable Timestamp:    Fri Sep 20 03:11:08 2024

Parallelism:
  Num Processors:          300
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:	   2
  Nodes:
    Total:                 1002001
    Local:                 3468
    Min/Max/Avg:           3173/3493/3340
  Elems:
    Total:                 1000000
    Local:                 3342
    Min/Max/Avg:           3278/3389/3333
  Num Subdomains:          1
  Num Partitions:          300
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                17034017
  Num Local DOFs:          58956
  Num Constrained DOFs:    34017
  Local Constrained DOFs:  0
  Variables:               { "gr0" "gr1" "gr2" "gr3" "gr4" ... "gr12" "gr13" "gr14" "gr15" "gr16" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                6004002
  Num Local DOFs:          20304
  Num Constrained DOFs:    4002
  Local Constrained DOFs:  0
  Variables:               "bnds" { "unique_grains" "var_indices" } "phi" { "err_bnds" "ind_bnds" }
  Finite Element Types:    "LAGRANGE" "MONOMIAL" "LAGRANGE" "MONOMIAL"
  Approximation Orders:    "FIRST" "CONSTANT" "FIRST" "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  TimeIntegrator:          BDF2
  Solver Mode:             Preconditioned JFNK
  PETSc Preconditioner:    hypre boomeramg

^[[31mLEGACY MODES ENABLED:^[[39m
 This application uses the legacy initial residual evaluation behavior. The legacy behavior performs an often times redundant residual evaluation before the solution modifying objects are executed prior to t$
^[[39m
Currently Executing
  Performing Initial Setup
    Computing User Objects
      Computing Polycrystal Initial Condition.                                           [^[[33m 15.03 s^[[39m] [^[[33m 1495 MB^[[39m]
      Finished Finalizing Polycrystal Initial Condition                                  [^[[33m  3.16 s^[[39m] [^[[33m 1854 MB^[[39m]
    Finished Computing User Objects                                                      [^[[33m 18.19 s^[[39m] [^[[33m 1854 MB^[[39m]
.
.
.
```
Regards

**Reply by GiudGiud** [reply]
Have you tried using more and less processors?
300 might not be optimal

**Reply by ashishdhole** [reply]
I have tried 240 before and 500 but I was just using the resources. I did not see any difference in the simulations time. Does it usually take this much time? I was told that this particular system is very small and should not take days to run it.

**Reply by GiudGiud** [reply]
@laagesen @dschwen any advice on improving phase field solve performances?

**Reply by GiudGiud** [reply]
Do you have a reference for how much time it should take?
Next steps if we want to improve the time is profiling.  We need to understand what takes time in this simulation
See this page for information on how to do this
https://mooseframework.inl.gov/application_development/profiling.html

Once we have a call graph we can take a look

**Reply by ashishdhole** [reply]
I did profiling. and got following 4 files. (Ran a small simulation with 4 cores on local pc, all the parameters kept same just the system is made smaller)

https://drive.google.com/drive/folders/1R9hSCuBlifq9ZwyqjmpVOPBdTPQE-FG0?usp=sharing

pprof takes very long time to even start
```
(moose) ashish@ashish-Inspiron-14-5430:~/projects/pfubc/Final_FP/polycrystals$ mpirun -np 4 pprof run1_0.prof 

```
it is stuck here. one forum says it is unmaintained package and hence takes time. Can you have a look at these files?

**Reply by GiudGiud** [reply]
I ll try next week. I think I need the executable as well. I m not sure I ll be able to parse pprof files without it

**Reply by GiudGiud** [reply]
you could try with google perf tools instead. That's definitely not unmaintained

**Comment by laagesen** [comment]
Hi, it is good that you are using GrainTracker, but you should be able to use fewer than 17 OPs. Try 8-10 for 2D. Also check your adaptivity setting to make sure you are not over-resolving the interfaces. You need about 3 elements through the interface between grains.

I would start from the example input file `moose/modules/phase_field/examples/grain_growth/grain_growth_2D_graintracker.i` in the example and _slowly_ make changes starting from this. I think this will be easier than trying to profile the code for your file. Try increasing the size and number of grains by a factor of 2 starting from this input file rather than trying to dramatically increasing the problem size and number of cores all at once. This will make it easier to identify issues.

**Reply by ashishdhole** [reply]
I am using following setup, 
other than op num do you think my choice of mesh points are okay. Is there a direct relation between the size of simulation, number of grains? 
```[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 250
  ny = 250
  xmin = 0
  xmax = 2000
  ymin = 0
  ymax = 2000
  uniform_refine = 2
[]
[GlobalParams]
  op_num = 17
  var_name_base = gr
[]

[Variables]
  [PolycrystalVariables]
    order = FIRST
    family = LAGRANGE
  []
[]
[Modules]
  [AdvanceFrictionPressureGG]
    g = phi
  []
[]

[UserObjects]
  [voronoi]
    type = PolycrystalVoronoi
    grain_num = 6400 # Number of grains
    rand_seed = 2878
    coloring_algorithm = jp
    int_width = 3.46
  []
  [grain_tracker]
    type = GrainTracker
    verbosity_level = 0
    compute_var_to_feature_map = true
  []
[]
```

**Comment by laagesen** [comment]
From what you have shown here, you'd have 1000 elements in each direction, starting from 250 elements and refining twice. That would give a mesh size of 2000 /  1000 = 2.0 in the x and y directions. This would be a little too coarse for interface width of 3.46. You would want at least 3 elements through the interface so probably at least 1 more level of refinement at the interfaces, adaptive meshing would work well to give you that. 

That said 6400 is A LOT of grains even for 2D. You are likely going to need a lot more cores. Best performance will be obtained at 20,000 DOF/core. It's very time consuming to troubleshoot performance and adaptivity for large simulations, which is why I strongly suggest to start with a much smaller problem and make sure your settings are working correctly before doing something this large.

**Reply by ashishdhole** [reply]
BAsed on the usage of cores, 1 core for 20000
my system has following DOFs
```
Nonlinear System:
  Num DOFs:                28774217
  Num Local DOFs:          82654
  Num Constrained DOFs:    44217
  Local Constrained DOFs:  0
  Variables:               { "gr0" "gr1" "gr2" "gr3" "gr4" ... "gr12" "gr13" "gr14" "gr15" "gr16" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                6762601
  Num Local DOFs:          18947
  Num Constrained DOFs:    2601
  Local Constrained DOFs:  0
  Variables:               "bnds" { "unique_grains" "var_indices" "phi" }
  Finite Element Types:    "LAGRANGE" "MONOMIAL"
  Approximation Orders:    "FIRST" "CONSTANT"
```
that means I have to use ~1438 cores to run this simulation. Is it correct? That is seriously high number of cores I need.

**Reply by laagesen** [reply]
Well you can reduce that by at least half if you use fewer order parameters

**Reply by ashishdhole** [reply]
I have tried that but 17 is the minimum OP I can get the simulation started. if I am using even finer mesh then the simulation terminates with out of memory issue.

**Reply by GiudGiud** [reply]
you dont "need" it. It s just the expected scalability limit

Having less is fine
The advice on reducing the order parameter is what you should look into

**Reply by ashishdhole** [reply]
ok I will see what changes I can make. Thank you

**Reply by laagesen** [reply]
Can you post an image of your initial microstructure? Images both with and without the mesh superimposed on top would be helpful. Trying to understand why you are needing such a large number of order parameters

**Reply by ashishdhole** [reply]
![image](https://github.com/user-attachments/assets/2ad3acd6-a6f5-473f-ab92-bbe7777d9ac1)

![image](https://github.com/user-attachments/assets/64ed571a-d75a-4de8-be67-93b845f6cfeb)

![image](https://github.com/user-attachments/assets/4dd7aa60-98a0-45dc-9fac-717950a4d217)

**Reply by laagesen** [reply]
As previously mentioned, you will want more elements across the interfaces, right now the interfaces are under-resolved and you are not going to properly resolve the physics of grain growth. The interfaces are also pretty wide relative to the size of the grains in some places, usually you want the grain size to be at least 10x larger than the interface width (although some variation in initial conditions makes this hard to achieve uniformly). Both of these factors could be part of the reason why you are needing more order parameters than usual.

Again I would highly suggest that you start with a much smaller problem and make sure all your settings and parameters are working properly before trying to run something this large.

**Reply by ashishdhole** [reply]
okay I get it now. that my mesh grids are not large enough. The number of grains are for a particular reason. The friction pressure model that I am working on need to have theses number of grains to see the pinning effect. So I will try to increase the resolution. I will get back to you with the update. thank you very much

**Reply by ashishdhole** [reply]
Hello,
One small doubt. If I use Adaptivity will the mesh adjust the resolution on its own?

I am using an Adaptivity block as follows 
```
[Adaptivity]
  initial_steps = 2
  max_h_level = 2
  marker = err_bnds
  [Markers]
    [err_bnds]
      type = ErrorFractionMarker
      coarsen = 0.3
      refine = 0.8
      indicator = ind_bnds
    []
  []
  [Indicators]
    [ind_bnds]
      type = GradientJumpIndicator
      variable = bnds
    []
  []
[]
```
Just curious. Thanks

**Reply by GiudGiud** [reply]
that is the intent with adaptivity, but it might take a little bit of tuning

**Reply by ashishdhole** [reply]
can you suggest something that suits this system? thanks

**Reply by GiudGiud** [reply]
there are inputs in the phase field module you can look at for using adaptivity and grain tracker iirc

**Comment by ashishdhole** [comment]
Hello, do you think GPU will help me do this large simulations? Are there any input files that I can look into? 
Thank you

**Reply by GiudGiud** [reply]
we are working on it. There is nothing ready right now for this

**Reply by ashishdhole** [reply]
ok. I think that would have helped me. I was looking at a paper (https://doi.org/10.1016/j.commatsci.2020.109992)
"The domain was divided into a grid of 2560 × 2560 × 2560 by regular [cubic lattices](https://www.sciencedirect.com/topics/earth-and-planetary-sciences/cubic-lattice) with a size of Δx = 1. The system consisted of 3,125,000 grains that are under periodic boundary conditions in all directions.
.
.
.
All simulations were carried out on the GPU-rich [supercomputer](https://www.sciencedirect.com/topics/earth-and-planetary-sciences/supercomputer) TSUBAME3.0 at the Tokyo Institute of Technology by utilizing our own parallel [GPU](https://www.sciencedirect.com/topics/engineering/graphics-processing-unit) code (https://www.sciencedirect.com/science/article/pii/S0927025620304833#b0330). The number of GPUs (NVIDIA [Tesla](https://www.sciencedirect.com/topics/engineering/tesla-roadster) P100) used for each simulation was 256."

That is a VERY LARGE system. 
Looking forward to the GPU based input parameters. 
Thank you
